airflow/providers_manager.py

```diff
@@ -46,42 +50,136 @@ class ProvidersManager:
-            except FileNotFoundError:
-                # This is OK - this is not a provider package
-                pass
-            except TypeError as e:
-                if "is not a package" not in str(e):
-                    log.warning("Error when loading 'provider.yaml' file from %s:%s}", module_info.name, e)
-                # Otherwise this is OK - this is likely a module
-            except Exception as e:  # noqa pylint: disable=broad-except
-                log.warning("Error when loading 'provider.yaml' file from %s:%s", module_info.name, e)
+            self._validator.validate(provider_info)
+            provider_info_package_name = provider_info['package-name']
+            if package_name != provider_info_package_name:
+                raise Exception(
+                    f"The package '{package_name}' from setuptools and "
+                    f"{provider_info_package_name} do not match. Please make sure they are"
+                    f"aligned"
+                )
+            if package_name not in self._provider_set:
+                self._provider_set.add(package_name)
+                self._provider_list.append((package_name, version, provider_info))
+            else:
+                log.warning(
+                    "The providers for package '%s' could not be registered because providers for that "
+                    "package name have already been registered",
+                    package_name,
+                )
+
+    def __find_all_airflow_builtin_providers_from_local_sources(self) -> None:
+        """
+        Finds all built-in airflow providers if airflow is run from the local sources.
+        It finds `provider.yaml` files for all such providers and registers the providers using those.
+
+        This 'provider.yaml' scanning takes precedence over scanning packages installed because
+        in case you have both sources and packages installed, the providers will be loaded from
+        the "airflow" sources rather than from the packages.
+        """
+        import airflow.providers
+
+        try:
+            for path in airflow.providers.__path__:
+                self.__add_provider_info_from_local_source_files_on_path(path)
+        except Exception as e:  # noqa pylint: disable=broad-except
+            log.warning("Error when loading 'provider.yaml' files from airflow sources: %s", e)
+
+    def __add_provider_info_from_local_source_files_on_path(self, path) -> None:
+        """
+        Finds all the provider.yaml files in the directory specified.
+        :param path: path where to look for provider.yaml files
+        """
+        root_path = path
+        for folder, _, files in os.walk(path):
+            for filename in fnmatch.filter(files, "provider.yaml"):
+                self.__add_provider_info_from_local_source_file(filename, folder, root_path)
+
+    def __add_provider_info_from_local_source_file(self, filename, folder, root_path) -> None:
```

Comment:
A signature change please:

```suggestion
    def __add_provider_info_from_local_source_file(self, filename, provider_dist_name) -> None: 
```
Where `filename` is the full path to the yaml file to load, and provider_dist_name is the name to register it as.
(dist_name not package_name, as `-` is not valid in packages).

For example, called like this:

```python
        for folder, _, files in os.walk(path):
            for filename in fnmatch.filter(files, "provider.yaml"):
                dist_name = "apache-airflow-providers" + folder[len(root_path) + 1 :].replace(os.sep, "-")
                self.__add_provider_info_from_local_source_file(os.path.join(folder, filename), dist_name) 
```

I think this change will make it easier to re-use this fn for plugins.

Reply:
Makes sense